### ISSUE TITLE
Karma test run fails frequently with 404s for bundled script files

### ISSUE THREAD

**Issue by sam-piper** [aurelia/cli#685]
**I'm submitting a bug report**

* **Library Version:**
aurelia-cli@0.30.1
karma@1.7.0
karma-chrome-launcher@2.2.0
karma-cli@1.0.1
karma-jasmine@1.1.0
karma-typescript-preprocessor@0.3.1

* **Operating System:**
Windows 10

* **Node Version:**
6.10.1

* **NPM Version:**
3.10.10

* **Browser:**
Chrome 60

* **Language:**
TypeScript 2.2.2

**Current behavior:**
We are experiencing major stability issues with the Karma web server when running tests using `au test --watch`.

Sometimes the tests run correctly for a test run, but very often the task fails with a whole slew of 404 warnings for bundled files, such as the following:

```
28 07 2017 12:22:01.601:INFO [launcher]: Launching browser Chrome with unlimited concurrency
28 07 2017 12:22:01.613:INFO [launcher]: Starting browser Chrome
28 07 2017 12:22:03.266:INFO [Chrome 60.0.3112 (Windows 10 0.0.0)]: Connected on socket VRL8rnxw6pOi57IrAAAA with id 187727
28 07 2017 12:22:03.676:WARN [web-server]: 404: /base/src/app.js
28 07 2017 12:22:03.678:WARN [web-server]: 404: /base/src/services/security/SecurityService.js
28 07 2017 12:22:03.678:WARN [web-server]: 404: /base/src/services/resources/ResourceService.js
28 07 2017 12:22:03.679:WARN [web-server]: 404: /base/src/model/EventNames.js
28 07 2017 12:22:03.679:WARN [web-server]: 404: /base/src/navigation/AuthorizeStep.js
28 07 2017 12:22:03.682:WARN [web-server]: 404: /base/src/model/observers/EntityGraphObserver.js
28 07 2017 12:22:03.683:WARN [web-server]: 404: /base/src/model/common-data/BankAccount.js
28 07 2017 12:22:03.684:WARN [web-server]: 404: /base/src/model/ObjectState.js
28 07 2017 12:22:03.684:WARN [web-server]: 404: /base/src/model/common-data/Address.js
Chrome 60.0.3112 (Windows 10 0.0.0) ERROR
  Uncaught Error: Script error for "src/app"
  http://requirejs.org/docs/errors.html#scripterror
  at wwwroot/scripts/vendor-bundle.js:4389
```

The exact files that result in 404s varies a lot as well, but they are always coming from `app-bundle.js`. The tests may run successfully to begin with, but if the underlying app code is touched then the next test run often breaks. Basically, it's flaky as hell - on some of our developers' boxes, they can barely get past this issue at all, so our developer workflow is broken in this area.

Here is our Karma configuration (very close to the base CLI file):

```
"use strict";
const path = require("path");
const project = require("./aurelia_project/aurelia.json");
const tsconfig = require("./tsconfig.json");

let testSrc = [
    { pattern: project.unitTestRunner.source, included: false },
    "test/aurelia-karma.js",
    "node_modules/jasmine-jquery/lib/jasmine-jquery.js",
    { pattern: "test/locales/*.json", watched: true, served: true, included: false },
    { pattern: "wwwroot/img/**/*", watched: false, served: true, included: false }  
];

let output = project.platform.output;
let appSrc = project.build.bundles.map(x => path.join(output, x.name));
let entryIndex = appSrc.indexOf(path.join(output, project.build.loader.configTarget));
let entryBundle = appSrc.splice(entryIndex, 1)[0];
let files = [entryBundle].concat(testSrc).concat(appSrc);

module.exports = function (config) {
    config.set({
        basePath: "",
        frameworks: [project.testFramework.id],
        files: files,
        proxies: {
            "/img/": "/base/wwwroot/img/"
        },
        exclude: [],
        preprocessors: {
            [project.unitTestRunner.source]: [project.transpiler.id]            
        },
        typescriptPreprocessor: {
            typescript: require("typescript"),
            options: tsconfig.compilerOptions
        },
        reporters: ["progress"],
        port: 9876,
        colors: true,
        logLevel: config.LOG_INFO,
        autoWatch: true,
        browsers: ["Chrome"],
        singleRun: false,
        // client.args must be a array of string.
        // Leave 'aurelia-root', project.paths.root in this order so we can find
        // the root of the aurelia project.
        client: {
            args: ["aurelia-root", project.paths.root]
        }
    });
};
```

We haven't made any changes to `/test/aurelia-karma.js`, or to the `au test` task, they are identical to current CLI versions for 0.30.1.

We really need some help in this area.

**Expected/desired behavior:**
Running tests should be stable for the Karma configuration, and never result in intermittent 404s to files that are bundled and loaded using RequireJS.

**Comment by sam-piper** [comment]
Just an additional note - if a test run succeeds, then we can make changes to any of the test spec files and they will run fine, it's only when the app files change that things break.

If the initial test run breaks, however, then we have to re-start the task.

**Comment by JeroenVinke** [comment]
@sam-piper thank you for opening the report. I haven't seen this issue before. It would help me if you could reproduce it in a new CLI application, and attach it as a ZIP to this issue. Then I can dig around and see what is happening

**Comment by sam-piper** [comment]
[CliTest.zip](https://github.com/aurelia/cli/files/1186282/CliTest.zip)

Here's a sample project that replicates the error. The error was replicated after I synched changes from our project to `/aurelia-project/tasks/process-css.ts`, and added in our SASS files. 

This is our customized CSS task:

```
import * as gulp from 'gulp';
import * as sourcemaps from 'gulp-sourcemaps';
import * as sass from 'gulp-sass';
import * as neat from 'node-neat';
import * as normalize from 'node-normalize-scss';
import * as sassGlob from 'gulp-sass-glob';
import * as autoprefixer from 'gulp-autoprefixer';
import * as project from '../aurelia.json';
import { build } from 'aurelia-cli';

export default function processCSS() {
    return gulp.src(project.cssProcessor.source)
        .pipe(sourcemaps.init())
        .pipe(sassGlob())
        .pipe(sass({ includePaths: neat.includePaths.concat(normalize.includePaths) }).on('error', sass.logError))
        .pipe(autoprefixer({
          browsers: ['last 3 versions'],
          cascade: false
        }))
        .pipe(build.bundle());
};
```

It's entirely possible that the use of SASS globbing and/or Autoprefixer is causing a side effect here, but I have no idea why it would cause this problem in Karma.

If you run `au test --watch` and start making changes to app.ts, you should start seeing the same error.

**Comment by WizzApp** [comment]
We are running Karma on our Jenkins (executing tests on Chrome, Firefox, IE and PhantomJs, just to be sure ;-) and we have the same issue. Strange thing is that it only seems to affect Chrome, the other browsers run fine with the same files. We also have SASS enabled here.

**Comment by JeroenVinke** [comment]
Thanks @sam-piper and @WizzApp. I've looked at this issue and found that if I use a minimal scss file (like one scss file with `body { background-color: red}`) I don't get this issue. It might not be related to SASS, but instead it might have to do with the size of the app-bundle. Weird thing is that when I refresh Karma in the browser the tests run again.  I haven't been able to get to the bottom of it yet, hopefully I can tell more in the next couple of days

**Comment by sam-piper** [comment]
Yeah I played around with the process-css task in the sample project to see if there were any specific elements causing problems.

I still get the error even after stripping out globbing, sourcemaps and Autoprefixer. It definitely seems to be a function of the amount of SASS code being handled, not the syntax or task configuration, as you've found as well.

On 2 Aug 2017, at 6:28 PM, Jeroen Vinke <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks @sam-piper<https://github.com/sam-piper> and @WizzApp<https://github.com/wizzapp>. I've looked at this issue and found that if I use a minimal scss file (like one scss file with body { background-color: red}) I don't get this issue. It might not be related to SASS, but instead it might have to do with the size of the app-bundle. Weird thing is that when I refresh Karma in the browser the tests run again. I haven't been able to get to the bottom of it yet, hopefully I can tell more in the next couple of days

**Comment by JeroenVinke** [comment]
I may have a lead. When I delay execution of the test files it seems to resolve the issue. Try and replace [this line](https://github.com/aurelia/cli/blob/master/lib/resources/test/require.aurelia-karma.js#L81) in aurelia-karma.js with `setTimeout(() => require(allTestFiles, window.__karma__.start), 2000)`.

If you can confirm that this helps I can look for a more decent solution. Let me know

**Comment by sam-piper** [comment]
That seems to resolve the problem completely. I don't even notice the 2 second timeout in the test console, looks like a race condition happening somewhere perhaps?

Thanks for this fix - happy to use this until you can find a better one.

**Comment by WizzApp** [comment]
Same here, after the change the tests run without issues.

**Comment by JeroenVinke** [comment]
Thanks @sam-piper and @WizzApp. Yes there seems to be a race condition happening somewhere between the loading of the app bundle and the start of the tests. I'll post here when I find a solution

**Comment by diegohb** [comment]
This fix did not work for me but i'm pretty sure it's because I'm using multiple bundles as well (related [to this](https://github.com/aurelia/cli/issues/203))... When i revert [this change](https://github.com/diegohb/aurelia-leaflet-cluster/commit/11a792cad8cbc1f4b1c0bf98cf1820769d89c70f) to split the bundles, i don't see any error running tests with AU TEST.

my repro code that shows the bug running AU TEST: https://github.com/diegohb/aurelia-leaflet-cluster/commit/ac15017d6b35432eb32fffa7b4417083703b47a4

The error stack trace:

> 
> Chrome 60.0.3112 (Windows 10 0.0.0) WARN: '%cUnhandled rejection Error: Fetch error: 404 Not Found
>   Instantiating http://localhost:9876/scripts/infra.vendor-bundle.js
>   Loading infra.vendor-bundle
>   Instantiating http://localhost:9876/jquery
>   Loading http://localhost:9876/app
>   Loading http://localhost:9876/base/src/app
>   Loading http://localhost:9876/base/test/unit/app.spec.js
>   Loading /base/test/unit/app.spec.js
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:10911
>     at Be (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:29826)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:28855
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:28832
>     at Ve.Ne (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:27839)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:3425
>     at h (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:3375)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:3105
>     at p (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:2933)
>     at Ve.f.(anonymous function) (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:38868)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:37134
>     at Ve.i.import (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:37106)
>     at Ve.import (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:45333)
>     at $e (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:29557)
>     at Ve.Ne (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:27821)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:3425
>     at h (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:3375)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:4390
>     at m (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:4178)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:5123
>     at <anonymous>
>     at b (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:5031)
>     at b (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:5485)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:39047
>     at Ve.f.(anonymous function) (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:38881)
>     at http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:37134
>     at <anonymous>
>     at Ve.i.import (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:37106)
> From previous event:
>     at Ve.i.import (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:37155)
>     at Ve.import (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:45333)
>     at t (http://localhost:9876/base/scripts/infra.au-bundle.js?81ea27d4adf8910301fc985616f337965fd87a7a:3788:41401)
>     at setTimeout (http://localhost:9876/base/test/aurelia-karma.js?2d2dcb7b71d474594d132e8787cfdb18ee81c977:43:22)', 'color: red'
> 
> 08 08 2017 16:54:24.331:WARN [Chrome 60.0.3112 (Windows 10 0.0.0)]: Disconnected (1 times), because no message in 10000 ms.
> 
> { uid: 11,
>   name: 'karma',
>   branch: false,
>   error: 1,
>   duration: [ 19, 747161702 ],
>   time: 1502225664858 }
> { uid: 0,
>   name: '<series>',
>   branch: true,
>   error: 1,
>   duration: [ 25, 649736261 ],
>   time: 1502225664860 }
> 1

**Comment by JeroenVinke** [comment]
@diegohb this looks to be a separate issue. Could you open a new issue? I'll try and look at the repro tomorrow.

**Comment by diegohb** [comment]
Created ticket https://github.com/aurelia/cli/issues/689

Thanks @JeroenVinke !

**Comment by sam-piper** [comment]
Just an update - we're still encountering the issue. It's less frequent, but it's still happening.

What's noticeable is that even if you increase that timeout to something like 60 seconds, there is no change in the behaviour of Karma, it starts as soon as it can.

**Comment by JeroenVinke** [comment]
I've just merged a PR that might fix this issue. Give it a try and let me know if it doesn't fix it

**Comment by sam-piper** [comment]
Great, thanks Jeroen - can you confirm the relevant commits for me? I'll merge them into our code and see if it works.

**Comment by JeroenVinke** [comment]
@sam-piper sure: https://github.com/aurelia/cli/commit/e773161b7ece0d5628a9b5d6364df5937be2bec9

**Comment by sam-piper** [comment]
So far this seems to be working - I've done about 30 test runs and no failures. Thanks for fixing this.

**Comment by WizzApp** [comment]
I'm still getting the `Firefox 54.0.0 (Windows 10 0.0.0) ERROR Error: Dynamic load not allowed: modules/machine/index.html at scripts/vendor-bundle.js:24832` errors ;-/
Seems like with Firefox Karma is executing the tests before the files are loaded properly ...

**Comment by JeroenVinke** [comment]
@wizzApp please open an issue with steps to reproduce, and we’ll have a look. Thanks

**Comment by WizzApp** [comment]
Will do if the issue persists.

FWIW and for others who stumble upon this issue: We set the concurrency down to 1 (where previously it was 4, running all browsers in parallel) as well as adding another short timeout between the require of `setup.js` and the rest of the test files in `aurelia-karma.js`. This fixed our problems with failing tests on Jenkins so all is well for now :-)

**Comment by JesseNewman19** [comment]
I'm still having this issue. Luckily @JeroenVinke's fix worked for me, but I can't say I'm a huge fan.

**Comment by WizzApp** [comment]
@JesseNewman19 Have a look at https://github.com/aurelia/cli/commit/e773161b7ece0d5628a9b5d6364df5937be2bec9
This contains the merged fix and does not use setTimeout anymore ...
